pkg/kubelet/kubelet.go

```diff
@@ -895,6 +888,64 @@ func NewMainKubelet(kubeCfg *kubeletconfiginternal.KubeletConfiguration,
 	return klet, nil
 }
 
+func finalizeTLSOptions(kubeCfg *kubeletconfiginternal.KubeletConfiguration, kubeDeps *Dependencies, klet *Kubelet, certDirectory string) error {
+	if kubeDeps.TLSOptions == nil {
+		return nil
+	}
+
+	// the ordering of the TLS config mutations below is important since both GetCertificate and GetConfigForClient can be set
+
+	if kubeCfg.ServerTLSBootstrap && utilfeature.DefaultFeatureGate.Enabled(features.RotateKubeletServerCertificate) {
+		var err error
+		klet.serverCertificateManager, err = kubeletcertificate.NewKubeletServerCertificateManager(klet.kubeClient, kubeCfg, klet.nodeName, klet.getLastObservedNodeAddresses, certDirectory)
+		if err != nil {
+			return fmt.Errorf("failed to initialize certificate manager: %v", err)
+		}
+		kubeDeps.TLSOptions.Config.GetCertificate = func(_ *tls.ClientHelloInfo) (*tls.Certificate, error) {
+			cert := klet.serverCertificateManager.Current()
+			if cert == nil {
+				return nil, fmt.Errorf("no serving certificate available for the kubelet")
+			}
+			return cert, nil
+		}
+	}
+
+	klet.clientCertificateCAContentProvider = kubeDeps.ClientCertificateCAContentProvider
+	if klet.clientCertificateCAContentProvider != nil {
+		// Populate PeerCertificates in requests, but don't reject connections without verified certificates
+		kubeDeps.TLSOptions.Config.ClientAuth = tls.RequestClientCert
+	}
+
+	if getCert, certFile, keyFile := kubeDeps.TLSOptions.Config.GetCertificate, kubeDeps.TLSOptions.CertFile, kubeDeps.TLSOptions.KeyFile; getCert == nil && len(certFile) > 0 && len(keyFile) > 0 {
+		var err error
+		klet.certKeyContentProvider, err = dynamiccertificates.NewDynamicServingContentFromFiles("kubelet-serving-cert", certFile, keyFile)
```

Comment:
I wonder would it be a little nicer if:
1. We had a `kubeDeps.ServingCertificateProvider`
2. The server rotation logic, instead of producing a GetCertificate function, produced one of these providers.

I know this would take some violence to the certificate manager, so maybe its not worthwhile in this PE, but it would be nice to have a consistent interface for 'give me a serving certificate'

Reply:
> it would be nice to have a consistent interface for 'give me a serving certificate'

The differences between the interfaces were large enough that I decided not to try to consolidate.  The TLS config logic is at least contained in one function now so future refactors should be easier.